### PR TITLE
Add link_selections function for automated linked selections

### PR DESCRIPTION
**Pull request by jonmmease**
## Overview
This PR adds a `holoviews.selection.link_selections` function that is an initial implementation of the automatic linked selection approach proposed in https://github.com/pyviz/holoviews/issues/3842.

This function can be used to transform HoloViews objects into new objects with linked selections enabled.

### Dependencies
This PR depends on the following open PRs, and will be rebased once these are merged.
 - https://github.com/pyviz/holoviews/pull/3924
 - https://github.com/pyviz/holoviews/pull/3927

## Example notebook
For examples, see the notebook at https://anaconda.org/jonmmease/link_selections_pr/notebook.

Please play around with this notebook to get a feel for the workflow and capabilities and report back on how things look @jbednar @philippjfr @jlstevens.

## Highlights
Linked `Points` plots with Bokeh backend:
points_layout

---
Linked `Points`, `Histogram`, `Scatter3d`, and `Table` layout with Plotly backend:
plotly_selections

---
Linked `Violin`, `Points` and `Bivariate` overlay, and `Distribution` layout with Bokeh backend:
bivariate_selections

---
Selections on datashaded elements
datashader_selection

---
Linked plots across notebook output cells, and updating selection color after display:
cross_cell_linking

## Extensions
There is a lot more functionality that can be built on this foundation. Here are some directions:

 - The `link_selections` function only stores the latest selection. But the `_base_link_selections` base class is more general, and other subclasses can be added for cross filtering, and glue-style multi-color brushing.
    - For cross-filtering, I really like the 4-color approach used in https://plotly.github.io/plotly.js-crossfilter.js/.  I personally find this easier to make sense of compared to the 2-color approach used in http://crossfilter.github.io/crossfilter/ where the "filtered" data is fully removed from the plot.
    - For multi-colored brushing, the inspiration is the glue project http://docs.glueviz.org/en/stable/getting_started/index.html?highlight=subsets#defining-subsets.  This would require a small Panel GUI interface for choosing the current subset color and the subset mode (union, intersection, or subtraction).
 - Add `Bounds` element to display the selected region after a selection is completed. I think this will largely be a requirement for the cross-filter mode, otherwise it would be hard to make sense of the union of the selected regions across plots.
 - Selections could be displayed on grid elements (e.g. vector field, raster) if we had an `element.where` operation (https://github.com/pyviz/holoviews/issues/3933).
 - For full hvplot compatibility, will need to review hvplot to make sure source dataset information flows through all of the produced elements. This happens for some, but not all, elements at the moment.
 - Investigate geoviews compatibility
 - Dask optimization of `dim` expressions and `histogram` operation to make selection with datashader + histogram elements with Dask DataFrames scale well.
 - Support passing full Panel layouts to `link_selections` and traverse them to find `HoloViews` panes. This way a Panel layout with HoloViews panes that use different backends could be linked together.
 - Add a new `BoundsPoly` stream that is analogous to `BoundsXY` but for a polygon rather than a box. Use this to support defining selection regions using lasso selection.

Let me know which of these you think are capabilities that we really should have for the initial release. Thanks!

### CODE REVIEW

**Comment by jbednar** [comment]
Fabulous!  Can we have all of those things? :-) It is hard to choose!

**Comment by jonmmease** [comment]
after rebasing on the `pipeline` support, it's also possible to automate linked selections through custom aggregation steps. e.g.

```python
bars = iris_ds.reindex(
    kdims='species_id', vdims=['sepal_length']
).to(hv.Bars).aggregate(function=np.sum)
link_selections(bars + points2)
```
aggregation_selection

**Comment by philippjfr** [comment]
>after rebasing on the pipeline support, it's also possible to automate linked selections through custom aggregation steps

Very, very neat.

**Comment by jbednar** [comment]
Cool!!

**Comment by johnzzzzzzz** [comment]
nice

**Comment by jonmmease** [comment]
All of the dependent PRs have been merged, so I squashed and rebased this PR on master.   Review/feedback appreciated!

File: holoviews/core/util.py [hunk 1]
```diff
@@ -1849,17 +1849,10 @@ def expand_grid_coords(dataset, dim):
     dataset into an ND-array matching the dimensionality of
     the dataset.
     """
-    irregular = [d.name for d in dataset.kdims
-                 if d is not dim and dataset.interface.irregular(dataset, d)]
-    if irregular:
```
Comment (philippjfr) [inline holoviews/core/util.py]: How come this is no longer needed?

**Comment by philippjfr** [comment]
Started reviewing but it looks to me like something went wrong during a merge or rebase and a number of preceding PRs are being reverted by this one.

**Comment by philippjfr** [comment]
Once that's fixed I'm fine with merging this. I'd like to see a user guide on using linked selections along with a few reference and gallery examples but that can happen in later PRs.

**Comment by jonmmease** [comment]
Thanks for catching that @philippjfr, I think it's straightened out now and I won't force push on this PR again in case you want to leave comments inline.  Thanks for taking a look!

**Comment by jbednar** [comment]
Oh, that explains the seemingly unrelated changes!  The diffs make much more sense now!

**Comment by philippjfr** [comment]
There appears to be some Python 2 issue.

**Comment by jonmmease** [comment]
Yeah, those errors are a bit confusing. I'll build a local Python 2 environment and see what I can figure out.

**Comment by philippjfr** [comment]
Very excited about this! Going to merge now.

**Comment by jonmmease** [comment]
Great! Thanks for tracking down the python 2 issue!

**Pull request merged**